### PR TITLE
Migrate rustc_mir_build diagnostics

### PR DESCRIPTION
**Pull request by mejrs**
Rebases https://github.com/rust-lang/rust/pull/100854

~~The remaining issue is how to better resolve https://github.com/rust-lang/rust/commit/72bea68af4ee2a41c44998916f6a789163f12e7d~~

~~The diagnostic macros seems to generate a broken diagnostic, and I couldn't figure out how to manually format the fluent message, so I hardcoded the format string for now. I'd like pointers to a better fix for this.~~

Also, I'm not 100% sure I didn't mess up a rebase somewhere 🙂 

r? @davidtwco

### CODE REVIEW

**Comment by davidtwco** [comment]
> The remaining issue is how to better resolve https://github.com/rust-lang/rust/commit/72bea68af4ee2a41c44998916f6a789163f12e7d
>
> The diagnostic macros seems to generate a broken diagnostic, and I couldn't figure out how to manually format the fluent message, so I hardcoded the format string for now. I'd like pointers to a better fix for this.

You should be able to fix this by using `#[subdiagnostic(eager)]` where this subdiagnostic is used.

**Comment by mejrs** [comment]
> You should be able to fix this by using `#[subdiagnostic(eager)]` where this subdiagnostic is used.

Ah that's good to know. I've fixed that and rebased.

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 1]
```diff
@@ -760,15 +676,17 @@ fn non_exhaustive_match<'p, 'tcx>(
...
-        pattern = "_".to_string();
+        cx.tcx.sess.emit_err(NonExhaustivePatternsTypeNotEmpty {
+            cx,
+            expr_span,
+            span: sp,
+            ty: scrut_ty,
+        });
+        return;
     } else {
+        // FIXME: migration of this diagnostic will require list support
```
Comment (davidtwco) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:689]: You can try using the list support added in #104047, maybe better as a follow-up so we can land this though.

**Review by davidtwco — APPROVED** [review]

**Comment by Manishearth** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/104790#issuecomment-1325615436

**Comment by mejrs** [comment]
Rebased

@rustbot ready

**Comment by mejrs** [comment]
I think this fixes the ice that was hit during rollup, but I'm not sure how to test it locally.

**Comment by mejrs** [comment]
I think this works on `dist-riscv64-linux` now - see https://github.com/rust-lang/rust/actions/runs/3721937031/jobs/6312424723

**Comment by Nilstrieb** [comment]
@bors r=davidtwco
Let's hope it really works now

**Pull request merged**